contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,361 @@
+    // logs unstructured request body
+    AuditRequestBody AuditLevel = "RequestBody"
+    // AuditResponseBody provides Request level of auditing, and additionally
+    // logs unstructured response body
+    AuditResponseBody AuditLevel = "ResponseBody"
+    // AuditStorageobject provides Response level, and additionally
+    // logs object before and after saving in storage
+    AuditStorageObject AuditLevel = "StorageObject"
+)
+
+type Event struct {
+    // AuditLevel at which event was generated
+    Level AuditLevel
+
+    // below fields are filled at HttpHeader level and higher:
+
+    // Unique audit ID, generated for each request
+    ID string
+    // Time the event reached the apiserver
+    Timestamp Timestamp
+    // Source IP, from where the request originates
+    SourceIP string
+    // HTTP method sent by the client
+    HttpMethod string
+    // Authentication method used to allow users access the cluster
+    AuthMethod string
+    // RequestURI is the Request-Line as sent by the client to a server
+    RequestURI string
+    // User information
+    User UserInfo
+    // Impersonate-User header value
+    AsUser string
+    // Impersonate-Group header value
+    AsGroups string
+    // Object reference this request is targeted at
+    Object ObjectReference
+    // Response status code are returned by the server
+    ResponseStatusCode int
+
+    // below fields are filled at RequestObject level and higher:
+
+    // RequestObject logged before admission
+    RequestObject runtime.Unstructured
+
+    // below fields are filled at StorageObject level and higher:
+
+    // Object value before modification (will be empty when creating new object)
+    OldObject runtime.Object
+    // Object value after modification (will be empty when removing object)
+    NewObject runtime.Object
+    // Difference between old and new object
+    Patch     []byte
+}
+```
+
+### Concurrency/Transaction Model for Events
+
+TODO: define (and maybe adapt interface) how synchronization of the writable audit event of a context takes place.
+
+### Policy
+
+The audit policy determines what audit event is generated for a given request. The policy is
+configured by the cluster administrator. Here is a sketch of the policy API:
+
+```go
+type Policy struct {
+    // Rules specify the audit Level a request should be recorded at.
+    // A request may match multiple rules, in which case the FIRST matching rule is used.
+    // The default audit level is None, but can be overridden by a catch-all rule at the end of the list.
+    Rules []PolicyRule
+
+    // Discussed under Filters section.
+    Filters []Filter
+}
+
+// Based off the RBAC PolicyRule
+type PolicyRule struct {
+    // Required. The Level that requests matching this rule are recorded at.
+    Level Level
+
+    // The users (by authenticated user name) this rule applies to.
+    // An empty list implies every user.
+    Users []string
+
```

Comment:
Yeah, I considered that but note that this is for the configuration, not the actual audit event. Assuming that the configuration is set up before the cluster is created, does including UID make any sense? Can you specify the UID of an object at creation?